src/appleseed/foundation/math/knn/knn_anyquery.h

```diff
@@ -0,0 +1,397 @@
+    {                                                                                           \
+        const NodeEntry tmp = node_queue[node_queue_size - (LhsIndex)];                         \
+        node_queue[node_queue_size - (LhsIndex)] = node_queue[node_queue_size - (RhsIndex)];    \
+        node_queue[node_queue_size - (RhsIndex)] = tmp;                                         \
+    }
+
+    const size_t NodeQueueSize = 128;
+    NodeEntry node_queue[NodeQueueSize];
+    size_t node_queue_size = 0;
+
+    // Start at the root.
+    node = nodes;
+
+    FOUNDATION_KNN_QUERY_STATS(++fetched_node_count);
+
+    while (node->is_interior())
+    {
+        const size_t split_dim = node->get_split_dim();
+        const ValueType split_abs = node->get_split_abs();
+        const ValueType split_dist = query_point[split_dim] - split_abs;
+
+        // Figure out which child node to follow (the one that contains the query point) and which one to push.
+        // Points on the split plane belong to the right child node.
+        const size_t follow_index = split_dist >= ValueType(0.0) ? 1 : 0;
+        const NodeType* APPLESEED_RESTRICT left_child_node = nodes + node->get_child_node_index();
+        const NodeType* APPLESEED_RESTRICT follow_node = left_child_node + follow_index;
+        const NodeType* APPLESEED_RESTRICT queue_node = left_child_node + 1 - follow_index;
+
+        FOUNDATION_KNN_QUERY_STATS(++fetched_node_count);
+
+        // Like in the initial step, we stop as soon as we reached a node with enough points.
+        if (follow_node->get_point_count() < IdealLeafSize)
+            break;
+
+        // Push the node that we don't visit now to the node stack.
+        const ValueType square_split_dist = square(split_dist);
+        if (square_split_dist < query_max_square_distance)
+        {
+            VectorType dvec(0.0);
+            dvec[split_dim] = split_dist;
+
+            // Push the node to the node stack.
+            assert(node_queue_size < NodeQueueSize);
+            node_queue[node_queue_size++] = NodeEntry(queue_node, dvec, square_split_dist);
+
+            // Order the top levels of the node stack.
+            if (node_queue_size >= 4)
+            {
+                ORDER_NODE_ENTRIES(4, 3);
+                ORDER_NODE_ENTRIES(2, 1);
+                ORDER_NODE_ENTRIES(4, 2);
+                ORDER_NODE_ENTRIES(3, 1);
+                ORDER_NODE_ENTRIES(3, 2);
+            }
+        }
+
+        // Continue with the child node.
+        node = follow_node;
+    }
+
+    //
+    // Step 4:
+    //
+    //   Visit the nodes in the node stack, from the closest to the farthest,
+    //   updating the stack with new nodes to visit later. We terminate as soon
+    //   as the closest node is farther than our current maximum search distance.
+    //
+
+    while (node_queue_size-- > 0)
+    {
+        const NodeEntry* top_entry = node_queue + node_queue_size;
```

Comment:
Again nitpicky. The `top_entry` is the lowest one in the tree you found and the last one added in the stack/queue/array. Maybe rename this variable, to indicate you're currently checking the smallest node (as close as possible to leafs) that's still a candidate.
This would make the code-snipplet better readable without reading through step 1-3.

Reply:
By "top" here I mean that it's the top item on the stack. Since the stack is only partially ordered, it's not guaranteed to be (and typically won't be) the closest one.